src/overlays/actors/ovl_En_Bal/z_en_bal.c

```diff
@@ -64,108 +165,999 @@ static DamageTable D_80A63CE0 = {
+    this->timer = 0;
+    this->actionFunc = EnBal_Talk;
+}
+
+void EnBal_Talk(EnBal* this, PlayState* play) {
+    Player* player;
+
+    player = GET_PLAYER(play);
+    player->stateFlags2 |= PLAYER_STATE2_100000;
+    this->isTalking = false;
+
+    switch (Message_GetState(&play->msgCtx)) {
+        case TEXT_STATE_NONE:
+        case TEXT_STATE_1:
+            break;
+        case TEXT_STATE_CLOSING:
+            break;
+        case TEXT_STATE_3:
+            if (this->lastMessageId != 0x1D10) {
+                this->isTalking = true;
+            }
+            break;
+        case TEXT_STATE_CHOICE:
+            EnBal_TryPurchaseMap(this, play);
+            break;
+        case TEXT_STATE_5:
+            EnBal_ProceedWithDialogue(this, play);
+            break;
+        case TEXT_STATE_DONE:
+            if (Message_ShouldAdvance(play)) {
+                EnBal_SetupGroundIdle(this);
+            }
+            break;
+        case TEXT_STATE_10:
+            if (Message_ShouldAdvance(play) && (this->lastMessageId == 0x1D08)) {
+                this->forceEyesShut = false;
+                this->eyeTexIndex = EN_BAL_EYETEX_OPEN;
+            }
+            break;
+        default:
+            break;
+    }
+
+    if (this->lastMessageId == 0x1D07) {
+        if ((this->skelAnime.curFrame > 29.0f) && (this->skelAnime.curFrame < 33.0f)) {
+            if (Animation_OnFrame(&this->skelAnime, 30.0f)) {
+                Actor_PlaySfx(&this->picto.actor, NA_SE_EV_CHINCLE_SPELL_EFFECT);
+            }
+            EnBal_ThrowMagicSparkles(this, play);
+        }
+        if (Animation_OnFrame(&this->skelAnime, 35.0f)) {
+            this->forceEyesShut = false;
+            this->eyeTexIndex = EN_BAL_EYETEX_OPEN;
+        }
+    }
+}
+
+void EnBal_UpdateShadow(EnBal* this) {
+    f32 homeY;
+    f32 worldY;
+    f32 shadowScale;
+
+    if ((this->actionFunc != EnBal_GroundIdle) && (this->actionFunc != EnBal_FloatIdle)) {
+        homeY = this->picto.actor.home.pos.y - this->picto.actor.floorHeight;
+        worldY = this->picto.actor.world.pos.y - this->picto.actor.floorHeight;
+        shadowScale = (homeY - worldY) / homeY;
+        shadowScale = CLAMP_MIN(shadowScale, 0.0f);
+        this->picto.actor.shape.shadowScale = 12.0f * shadowScale;
+        this->picto.actor.shape.shadowAlpha = 255.0f * shadowScale;
+    }
+}
+
+void EnBal_SetRecognizedPlayerForm(void) {
+    if (gSaveContext.save.playerForm != PLAYER_FORM_GORON) {
+        if (gSaveContext.save.playerForm != PLAYER_FORM_ZORA) {
+            if (gSaveContext.save.playerForm != PLAYER_FORM_DEKU) {
+                if (gSaveContext.save.playerForm == PLAYER_FORM_HUMAN) {
+                    SET_WEEKEVENTREG(WEEKEVENTREG_64_08);
+                    SET_WEEKEVENTREG(WEEKEVENTREG_64_10);
+                }
+            } else {
+                SET_WEEKEVENTREG(WEEKEVENTREG_64_08);
+            }
+        }
+    } else {
+        SET_WEEKEVENTREG(WEEKEVENTREG_64_10);
+    }
```

Comment:
Hmm this looks like a switch, where the cases go human, deku, zora, and than goron.